controller/Review.php

```diff
@@ -35,36 +44,98 @@
+     * @throws LtiInvalidLaunchDataException
      * @throws LtiVariableMissingException
      * @throws common_exception_Error
+     * @throws common_exception_NotFound
      */
-    public function index()
+    public function index(): void
     {
-        // TODO: Move logic from controller to service with test coverage
-
         $launchData = LtiService::singleton()->getLtiSession()->getLaunchData();
 
-        /** @var LtiLaunchDataService $launchDataService */
-        $launchDataService = $this->getServiceLocator()->get(LtiLaunchDataService::SERVICE_ID);
+        /** @var DeliveryExecutionFinderService $finder */
+        $finder = $this->getServiceLocator()->get(DeliveryExecutionFinderService::SERVICE_ID);
 
-        /** @var LtiResultAliasStorage $ltiResultIdStorage */
-        $ltiResultIdStorage = $this->getServiceLocator()->get(LtiResultAliasStorage::SERVICE_ID);
+        $execution = $finder->findDeliveryExecution($launchData);
+        $delivery = $execution->getDelivery();
 
+        $data = [
+            'execution' => $execution->getUri(),
+            'delivery'  => $delivery->getUri(),
+        ];
+
+        $this->composeView('delegated-view', $data, 'pages/index.tpl', 'tao');
+    }
 
-        $resultIdentifier = $launchData->hasVariable('lis_result_sourcedid')
-            ? $launchData->getVariable('lis_result_sourcedid')
-            : $launchDataService->findDeliveryExecutionFromLaunchData($launchData);
+    /**
+     * @throws common_Exception
+     */
+    public function init(): void
+    {
+        /** @var QtiRunnerInitDataBuilderFactory $dataBuilder */
+        $dataBuilder = $this->getServiceLocator()->get(QtiRunnerInitDataBuilderFactory::SERVICE_ID);
 
-        $deliveryExecutionId = $ltiResultIdStorage->getDeliveryExecutionId($resultIdentifier);
+        $params = $this->getPsrRequest()->getQueryParams();
 
-        if ($deliveryExecutionId !== null) {
-            $params['delivery_execution'] = $deliveryExecutionId;
+        if (isset($params['serviceCallId'])) {
+            $data = $dataBuilder->create()->build($params['serviceCallId']);
         }
 
-        $this->setClientRoute(_url('index', 'Review', 'taoReview', $params ?? []));
-        $this->composeView('delegated-view', null, 'pages/index.tpl', 'tao');
+        $this->returnJson($data ?? []);
+    }
+
+    /**
+     * Provides the definition data and the state for a particular item
+     */
+    public function getItem(): void
+    {
+        $params = $this->getPsrRequest()->getQueryParams();
+
+        $deliveryExecutionId = $params['serviceCallId'];
+        $itemDefinition = $params['itemUri'];
+
+        /** @var DeliveryExecutionManagerService $des */
+        $deManagerService = $this->getServiceLocator()->get(DeliveryExecutionManagerService::SERVICE_ID);
+        $execution = $deManagerService->getDeliveryExecutionById($deliveryExecutionId);
+        $delivery = $execution->getDelivery();
+
+        $itemPreviewer = new ItemPreviewer();
```

Comment:
ItemPreviewer should be an immutable object which can be instantiated through a builder object